odoo/addons/base/models/ir_cron.py

```diff
@@ -280,73 +291,108 @@ def _acquire_one_job(cls, cr, job_ids):
             raise
         return cr.dictfetchone()
 
-    @classmethod
-    def _process_job(cls, db, cron_cr, job):
-        """ Execute a cron job and re-schedule a call for later. """
+    def _notify_admin(self, message):
+        _logger.warning(message)
 
-        # Compute how many calls were missed and at what time we should
-        # recall the cron next. In the example bellow, we fake a cron
-        # with an interval of 30 (starting at 0) that was last executed
-        # at 15 and that is executed again at 135.
-        #
-        #    0          60          120         180
-        #  --|-----|-----|-----|-----|-----|-----|----> time
-        #    1     2*    *     *     *  3  4
-        #
-        # 1: lastcall, the last time the cron was executed
-        # 2: past_nextcall, the cron nextcall as seen from lastcall
-        # *: missed_call, a total of 4 calls are missing
-        # 3: now
-        # 4: future_nextcall, the cron nextcall as seen from now
+    def _notify_admin_deactivation(self, cron_name, cron_id, action_id):
+        self._notify_admin(_('The cron %s with id %s call from action %s has been deactivated after failing 5 consecutive times.\n'
+                             'Verify the cron logs for more information.', cron_name, cron_id, action_id))
 
+    @classmethod
+    def _execute_job(cls, job):
+        """ Execute job ir_action up to MAX_BATCH_PER_CRON_JOB as long as it's not completed. """
         with cls.pool.cursor() as job_cr:
-            lastcall = fields.Datetime.to_datetime(job['lastcall'])
-            interval = _intervalTypes[job['interval_type']](job['interval_number'])
-            env = api.Environment(job_cr, job['user_id'], {'lastcall': lastcall})
+            env = api.Environment(job_cr, job['user_id'], {'lastcall': job['lastcall'], 'cron_id': job['id']})
             ir_cron = env[cls._name]
+            completed = True
+            for i in range(MAX_BATCH_PER_CRON_JOB):
+                progress = ir_cron._create_progress(job['id'], timed_out=True)
+                job_cr.commit()
+
+                ir_cron_w_context = ir_cron.with_context(ir_cron_progress_id=progress.id)
+                try:
+                    (done, remaining) = ir_cron_w_context._callback(
+                        job['cron_name'], job['ir_actions_server_id'], job['id']
+                    )
+                except Exception:
+                    failed = not progress.done  # If some progress has been committed, we do not consider it a failure
+                else:
+                    failed = False
+                    ir_cron_w_context._notify_progress(done, remaining)
+                finally:
+                    progress.timed_out = False
+                    job_cr.commit()
+
+                completed = not progress.remaining
+                if completed:
```

Comment:
Explicit it better than implicit. You told me that in case of a crash, the transaction would rollback and any change to the `progress` record would be lost, making `progess.remaining == 0` thus exiting the loop here. But in case the developers call `_notify_progress` themselves at the very beginning of their cron to notify how many records there are to process but didn't processed any yet, our assumption fails:

```py
records = self.search(...)
self.env['ir.cron']._notify_progress(done=0, remaining=len(records))
self.env.cr.commit()
for done, record in enumerate(records, 1):
    ...
    self.env['ir.cron']._notify_progress(done=done, remaining=len(records) - done)
    self.env.cr.commit()
```

So I suggest we make it explicit:
```suggestion
                if completed or failed:
```